src/Materials/PBR/pbrBaseMaterial.ts

```diff
@@ -975,13 +928,18 @@ export abstract class PBRBaseMaterial extends PushMaterial {
                 this._renderTargets.push(<RenderTargetTexture>this._reflectionTexture);
             }
 
-            this.subSurface.fillRenderTargetTextures(this._renderTargets);
+            Material.OnEventObservable.notifyObservers(
+                this,
+                MaterialEvent.FillRenderTargetTextures,
+                undefined,
+                undefined,
+                { renderTargets: this._renderTargets } as MaterialEventInfoFillRenderTargetTextures

```

Comment:
This creates GC

Reply:
@sebavan we were talking about this earlier:

> >  BTW, any specific reason to make Material._EventInfo a static private variable, instead of using a local variable? It's not reentrant.
> I did it static to avoid creating a new object each time we need to notify the observers. Dealing with reentrancy would mean to create an object each time we notify and I fear it would be bad for perf/GC... I don't know what @sebavan things about this? Should we document the fact that the material events are not reentrant and one should not implement an observer that would in turn would lead to sending notifications to another/others observers or should we handle it in the proper way?

What do you suggest? I think a static field in Material, shared by everything, is not a great idea. Perhaps a non-static field on the material instance, combining all the `MaterialEventInfo*` interfaces would be better? But it might keep some stale data that won't be GC'ed.

Reply:
Here's what I wrote about this above:
> I did it static to avoid creating a new object each time we need to notify the observers. Dealing with reentrancy would mean to create an object each time we notify and I fear it would be bad for perf/GC... I don't know what @sebavan things about this? Should we document the fact that the material events are not reentrant and one should not implement an observer that would in turn would lead to sending notifications to another/others observers or should we handle it in the proper way?

I think we should simply use a single event info object and reuse it each time we notifiy the observers as I did at first and document it is not reentrant and wait to see if it's a problem in the future.

Reply:
> Perhaps a non-static field on the material instance, combining all the MaterialEventInfo* interfaces would be better?

That won't help with the reentrancy problem. As javascript is mono-threaded, having a static field will work the same as having a specific instance per material, that's why I suggested using a static to save some memory. But using a single non-static property is fine by me too.

Reply:
I am not sure to understand. Let s say each material instance is having one observable instance instead of the global one.

Then the material keeps a private EventInfo field and use it on the events being fired. This way we should not have any problems. 

Any issues we can document would in any case be better than GC :-)

Reply:
If we are using a single private eventInfo field per material with have a possible reentrancy problem. Suppose an event is raised by a method of the material M and an observer of this event is calling a method of the same material M which itself raises an event: that will reuse the same eventInfo instance (because there's a single one for the material) than the one used for the first notification, which may be a problem (and which will be a problem if both events are the same).

Regarding not using a static observable but an instance per material is probably good: for the material plugin manager a static instance is enough because it attaches to all notifications of all materials. But if some user code want to attach to events of a single material instance it's probably better that the material has its own observable instance.

Reply:
Yup totally agree. 

Let s document the reentrancy and keep one instance of each eventInfo type limiting the reentrancy to the same event recursively firing ?

And lets make the handler an instance thingy :-)